Lookup show more

Blackbaud-TrevorBurch opened the pull request.

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

Blackbaud-SteveBrush commented on src/app/public/modules/lookup/lookup-show-more-modal.component.ts:
```diff
@@ -0,0 +1,331 @@
+   * @internal
+   * Fires when users select the "Add" button
+   */
+  @Output()
```
If this is truly internal-only, I would go with an observable and remove the `@Output` decorator.

Blackbaud-TrevorBurch replied:
This somewhat has changed in theory due to the split up of the modal work so I'm marking this as resolved.

Blackbaud-SteveBrush commented on src/app/public/modules/lookup/lookup-show-more-modal.component.ts:
```diff
@@ -0,0 +1,331 @@
+  providers: [SkyDataManagerService],
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class SkyLookupShowMoreModalComponent implements OnDestroy, OnInit {
```
Do we want this class showing up in the docs? Need an `@internal` JSDoc?

Blackbaud-TrevorBurch replied:
Not relevant now due to the modal work split up.

Blackbaud-SteveBrush commented on src/app/public/modules/lookup/lookup-show-more-modal.component.ts:
```diff
@@ -0,0 +1,331 @@
+  };
+
+  private itemIndex: number = 0;
+  private ngUnsubscribe = new Subject();
```
`Subject<void>()`

Blackbaud-TrevorBurch replied:
Fixed in lookup-modals PR

Blackbaud-SteveBrush commented on src/app/public/modules/lookup/lookup-show-more-modal.component.ts:
```diff
@@ -0,0 +1,331 @@
+      }
+    );
+
+    setTimeout(() => {
```
Mind adding a comment above this `setTimeout` explaining why it's needed?

Blackbaud-TrevorBurch replied:
I actually dug into this and was able to get around it by moving this logic from `ngOnInit` to `ngAfterViewInit`. This is done in the `lookup-modals` PR

Blackbaud-SteveBrush commented on src/app/public/modules/lookup/lookup-show-more-modal.component.ts:
```diff
@@ -0,0 +1,331 @@
+          debounceTime(250),
+          takeUntil(this.ngUnsubscribe)
+        ).subscribe(state => {
+        if (this.dataState.searchText !== state.searchText) {
```
Missing tab?

Blackbaud-TrevorBurch replied:
Fixed in `lookup-modals` PR

Blackbaud-SteveBrush commented on src/app/public/modules/lookup/lookup-show-more-modal.component.ts:
```diff
@@ -0,0 +1,331 @@
+        };
+      });
+      this.items.forEach(item => {
+        if ((this.isEquivalent(this.context.initialValue, item.value)) ||
```
The logic in this `if` block is a bit difficult to parse out, personally. What if we broke it out into separate statements to make it easier to read?

Blackbaud-TrevorBurch replied:
I split this out a little more in the `lookup-modals` PR.

Blackbaud-SteveBrush commented on src/app/public/modules/lookup/lookup.module.ts:
```diff
@@ -1,25 +1,79 @@
+    SkyRepeaterModule,
+    SkyThemeModule,
+    SkyTokensModule,
+    SkyViewkeeperModule
```
We're adding quite a few modules to Lookup's dependencies, including some bulky modules such as Data Manager and Repeater. Might be something to discuss over the phone, just to confirm that's the direction we want to go in.

Blackbaud-TrevorBurch replied:
We split things up for this with a new modal provider system.

Blackbaud-AlexKingman reviewed (commented):
Ran into some strange keyboard controls...
1) Go to the visual test
2) tab into the input box
3) Type "a" and wait for results to show
4) Hit the down key until you get to the last item in the list
5) Keep hitting the down key to see the selection disappear. Where did it go?

Blackbaud-TrevorBurch commented on src/app/public/modules/lookup/fixtures/lookup.component.fixture.html:
```diff
@@ -13,10 +13,29 @@
+  #customShowMoreTemplate
+>
+  {{ item.name }} - {{ item.birthDate }}
+</ng-template>ks
```
🤦 - fixed

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

Blackbaud-TrevorBurch commented on src/app/public/modules/autocomplete/autocomplete.component.html:
```diff
@@ -25,34 +25,53 @@
+          </ng-container>
+        </div>
+      </ng-container>
+      <div *ngIf="!(showAddButton || showMoreButton) && searchResults.length === 0"
```
I actually removed the complex logic here as it was redundant.

Blackbaud-TrevorBurch commented on src/app/public/modules/autocomplete/autocomplete.component.html:
```diff
@@ -25,34 +25,53 @@
     #resultsRef
   >
-    <div *ngIf="searchText"
+    <div *ngIf="searchText && (!(showAddButton || showMoreButton) || searchResults.length > 0)"
```
I've cleaned this up by adding a getter for `showActionsArea` that abstracts the `or` statement here. The rest is needed specifically here and isn't a great candidate to abstract.

Blackbaud-TrevorBurch commented on src/app/public/modules/autocomplete/autocomplete.component.html:
```diff
@@ -25,34 +25,53 @@
+          Show matches ({{searchResults.length}})
+        </ng-container>
+      </button>
+      <div *ngIf="(showAddButton || showMoreButton) && searchText && searchResults.length === 0"
```
I actually was able to remove the `showActionsArea` logic here as it was redundant due to a parent check already in place.

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

Blackbaud-SteveBrush commented on src/app/public/modules/autocomplete/types/autocomplete-default-search-function-options.ts:
```diff
@@ -8,5 +8,4 @@ import {
 export interface SkyAutocompleteDefaultSearchFunctionOptions {
   propertiesToSearch?: string[];
   searchFilters?: SkyAutocompleteSearchFunctionFilter[];
-  searchResultsLimit?: number;
```
This type is exported via the public API. Possible to avoid removing this value?

Blackbaud-TrevorBurch replied:
Ya. See my comment on the default function.

Blackbaud-TrevorBurch replied:
Added back

Blackbaud-SteveBrush commented on src/app/public/modules/lookup/lookup-show-more-modal.component.html:
```diff
@@ -0,0 +1,84 @@
+<sky-modal>
+  <sky-modal-header>
+    {{getModalTitle()}}
```
Using methods inside component templates [can cause performance problems](https://medium.com/showpad-engineering/why-you-should-never-use-function-calls-in-angular-template-expressions-e1a50f9c0496). Possible to use a static `modalTitle` property instead?

Blackbaud-TrevorBurch replied:
I actually just moved this logic to the template. I'm glad you caught this though because when I looked I realized I forgot to add the i18n strings for this title. I think in this case that the template makes more sense for this logic as having it in the component would make us deal with the i18n service which is much more gross than the pipe in the template.

Blackbaud-SteveBrush replied:
Your solution works well.

Blackbaud-TrevorBurch commented on src/app/public/modules/lookup/lookup.component.ts:
```diff
@@ -129,6 +145,34 @@ export class SkyLookupComponent
+    this._showMoreButton = value;
+
+    if (value && !this.modalProvider) {
+      this.modalProvider = this.modalProviderService.getModalForType('lookup-show-more');
```
Uhhhh..... this doesn't exist anymore. I think you might have been looking at an old version of the code here. This went away when we reverted away from the `lookup-modal` library stuff.

Blackbaud-SteveBrush replied:
Yep, this was an older comment from a prior commit; forgot to delete it!

Blackbaud-TrevorBurch commented on src/app/public/modules/autocomplete/autocomplete.component.ts:
```diff
@@ -213,9 +222,16 @@ export class SkyAutocompleteComponent
+   * Shows the Show more button in the actions bar
+   */
+  @Input()
+  public showMoreButton: boolean = false;
```
I'll touch base with you about this one.... this is a case where I wish I had named the `showAddButton` something different in the previous PR but it has already been released. You are correct in that I could rename this in `autocomplete` but I can't rename the `showAddButton` that exists on lookup itself because it is not internal in that case. I'll ping you about this.

Blackbaud-TrevorBurch commented on src/app/public/modules/autocomplete/autocomplete-default-search-function.ts:
```diff
@@ -44,15 +44,6 @@ export function skyAutocompleteDefaultSearchFunction(
-
-      if (limitReached) {
-        return results;
-      }
```
Ya. That is probably a good catch. The background here is that this was causing issues where we were limiting the search results being returned instead of limiting what was being shown from those results. This caused issues when we add a button that should be showing _all_ search results as we didn't have them. 

I look at this though and we can probably deprecate the `searchResultsLimit` and just change the default to not exist instead of defaulting to 5 which is what I believe it was in the past.

Blackbaud-TrevorBurch replied:
I added this back but did not add back where we were using this internally as we are handling this in the autocomplete template instead of this function now. This is because we need to have all search results but only show some of them. 

I'm not sure why we are exporting the options in the public API as it isn't documented, but since it is exported it is best that we don't remove a property.

Blackbaud-SteveBrush replied:
I’d rather do one of two things: 1) Remove the default function from our public API and make the necessary changes (since it's internal, and an audit of the functions usage turned up zero results), or 2) Keep it in the API and write unit tests for the existing “limit” logic. I’m personally okay with option 1 if you don’t see any red flags.

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

Blackbaud-TrevorBurch commented on src/app/public/modules/lookup/types/lookup-show-more-context.ts:
```diff
@@ -0,0 +1,21 @@
+ */
+ export class SkyLookupShowMoreContext {
+  public items: any[];
+  public descriptorProperty: any;
```
Good catch. Fixed

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)

Blackbaud-SteveBrush commented on src/app/public/modules/lookup/types/lookup-show-more-context.ts:
```diff
@@ -0,0 +1,21 @@
+ * @internal
+ * Context for the show more modal. These values are provided by the lookup component.
+ */
+ export class SkyLookupShowMoreContext {
```
Let's call this `SkyLookupShowMorePickerContext` just to be more specific.

Blackbaud-TrevorBurch replied:
I actually went with `SkyLookupShowMoreNativePickerContext` to distinguish that this is only used for the native picker as we already had a `SkyLookupShowMoreCustomPickerContext`

Blackbaud-SteveBrush reviewed (commented):
For all the JsDocs, I would replace the word "modal" with "picker" to keep it ambiguous, and to match the input names we're adding.

Blackbaud-TrevorBurch commented on src/app/public/modules/autocomplete/autocomplete.component.ts:
```diff
@@ -127,6 +127,13 @@ export class SkyAutocompleteComponent
 
+  /**
+   * @internal
+   * Shows the Show more button in the actions bar
```
I had actually updated this in lookup but for to update it here. Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/lookup/lookup.component.ts:
```diff
@@ -128,6 +154,11 @@ export class SkyLookupComponent
   @Input()
   public showAddButton: boolean = false;
+  /**
+   * Specifies the configuration options for the show more modal.
```
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/lookup/types/lookup-show-more-config.ts:
```diff
@@ -0,0 +1,22 @@
+  /**
+   * Specifies the configuration for the built in UI when users select the show more button.
+   */
+  defaultPickerConfig?: SkyLookupShowMoreDefaultPickerConfig;
```
Done. I also renamed the actual config object.

Blackbaud-TrevorBurch commented:
@Blackbaud-SteveBrush - ready for you again. I did leave the autocomplete default search options in the public API as I did find 4 SPAs that were referencing it (some with code copied partially from our default function). I removed the code in our default function that looks for `searchResultLimit` and then deprecated the result limit in the public options type.

I also updated JS Docs to use `picker` instead of `modal` everywhere.

blackbaud-ado commented:
Preview build at [https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more](https://developer.blackbaud.com/skyux-lookup?_sp.spa=pr-lookup-show-more)
